### ISSUE TITLE
compile error on Linux kernel 4.15.0

### ISSUE THREAD

**Issue by Tiryoh** [rt-net/RaspberryPiMouse#8]
a compile error occurred on Linux kernel 4.15.0 (Ubuntu Bionic)

```
ubuntu@ubuntu:~$ ~/RaspberryPiMouse/utils/build_install.ubuntu14.bash
#!/bin/bash -vxe

dir=$(dirname $0)/../
++ dirname /home/ubuntu/RaspberryPiMouse/utils/build_install.ubuntu14.bash
+ dir=/home/ubuntu/RaspberryPiMouse/utils/../

cd $dir/src/drivers/
+ cd /home/ubuntu/RaspberryPiMouse/utils/..//src/drivers/
rm Makefile
+ rm Makefile
ln -s Makefile.ubuntu14 Makefile
+ ln -s Makefile.ubuntu14 Makefile
make clean
+ make clean
make -C /usr/src/linux-headers-`uname -r` M=`pwd` V=1 clean
make[1]: Entering directory '/usr/src/linux-headers-4.15.0-1010-raspi2'
make -f ./scripts/Makefile.clean obj=/home/ubuntu/RaspberryPiMouse/src/drivers
  rm -rf /home/ubuntu/RaspberryPiMouse/src/drivers/.tmp_versions
  rm -f /home/ubuntu/RaspberryPiMouse/src/drivers/Module.symvers
make[1]: Leaving directory '/usr/src/linux-headers-4.15.0-1010-raspi2'
make
+ make
make -C /usr/src/linux-headers-`uname -r` M=`pwd` V=1 modules
make[1]: Entering directory '/usr/src/linux-headers-4.15.0-1010-raspi2'
test -e include/generated/autoconf.h -a -e include/config/auto.conf || (		\
echo >&2;							\
echo >&2 "  ERROR: Kernel configuration is invalid.";		\
echo >&2 "         include/generated/autoconf.h or include/config/auto.conf are missing.";\
echo >&2 "         Run 'make oldconfig && make prepare' on kernel src to fix it.";	\
echo >&2 ;							\
/bin/false)
mkdir -p /home/ubuntu/RaspberryPiMouse/src/drivers/.tmp_versions ; rm -f /home/ubuntu/RaspberryPiMouse/src/drivers/.tmp_versions/*
make -f ./scripts/Makefile.build obj=/home/ubuntu/RaspberryPiMouse/src/drivers
  gcc -Wp,-MD,/home/ubuntu/RaspberryPiMouse/src/drivers/.rtmouse.o.d  -nostdinc -isystem /usr/lib/gcc/arm-linux-gnueabihf/7/include  -I./arch/arm/include -I./arch/arm/include/generated  -I./include -I./arch/arm/include/uapi -I./arch/arm/include/generated/uapi -I./include/uapi -I./include/generated/uapi -include ./include/linux/kconfig.h -Iubuntu/include  -D__KERNEL__ -mlittle-endian -Wall -Wundef -Wstrict-prototypes -Wno-trigraphs -fno-strict-aliasing -fno-common -fshort-wchar -Werror-implicit-function-declaration -Wno-format-security -std=gnu89 -fno-PIE -fno-dwarf2-cfi-asm -fno-omit-frame-pointer -mapcs -mno-sched-prolog -fno-ipa-sra -mabi=aapcs-linux -mno-thumb-interwork -mfpu=vfp -funwind-tables -marm -D__LINUX_ARM_ARCH__=7 -march=armv7-a -msoft-float -Uarm -fno-delete-null-pointer-checks -Wno-frame-address -Wno-format-truncation -Wno-format-overflow -Wno-int-in-bool-context -O2 --param=allow-store-data-races=0 -DCC_HAVE_ASM_GOTO -Wframe-larger-than=1024 -fstack-protector-strong -Wno-unused-but-set-variable -Wno-unused-const-variable -fno-omit-frame-pointer -fno-optimize-sibling-calls -fno-var-tracking-assignments -pg -Wdeclaration-after-statement -Wno-pointer-sign -fno-strict-overflow -fno-merge-all-constants -fmerge-constants -fno-stack-check -fconserve-stack -Werror=implicit-int -Werror=strict-prototypes -Werror=date-time -Werror=incompatible-pointer-types -Werror=designated-init  -DMODULE  -DKBUILD_BASENAME='"rtmouse"'  -DKBUILD_MODNAME='"rtmouse"' -c -o /home/ubuntu/RaspberryPiMouse/src/drivers/.tmp_rtmouse.o /home/ubuntu/RaspberryPiMouse/src/drivers/rtmouse.c
/home/ubuntu/RaspberryPiMouse/src/drivers/rtmouse.c: In function ‘sw_read’:
/home/ubuntu/RaspberryPiMouse/src/drivers/rtmouse.c:566:6: error: implicit declaration of function ‘copy_to_user’; did you mean ‘raw_copy_to_user’? [-Werror=implicit-function-declaration]
  if (copy_to_user((void *)buf, &rw_buf, count)) {
      ^~~~~~~~~~~~
      raw_copy_to_user
/home/ubuntu/RaspberryPiMouse/src/drivers/rtmouse.c: In function ‘parseFreq’:
/home/ubuntu/RaspberryPiMouse/src/drivers/rtmouse.c:816:7: error: implicit declaration of function ‘copy_from_user’; did you mean ‘raw_copy_from_user’? [-Werror=implicit-function-declaration]
   if (copy_from_user(&cval, buf + i, sizeof(char))) {
       ^~~~~~~~~~~~~~
       raw_copy_from_user
cc1: some warnings being treated as errors
scripts/Makefile.build:339: recipe for target '/home/ubuntu/RaspberryPiMouse/src/drivers/rtmouse.o' failed
make[2]: *** [/home/ubuntu/RaspberryPiMouse/src/drivers/rtmouse.o] Error 1
Makefile:1552: recipe for target '_module_/home/ubuntu/RaspberryPiMouse/src/drivers' failed
make[1]: *** [_module_/home/ubuntu/RaspberryPiMouse/src/drivers] Error 2
make[1]: Leaving directory '/usr/src/linux-headers-4.15.0-1010-raspi2'
Makefile:7: recipe for target 'rtmouse.ko' failed
make: *** [rtmouse.ko] Error 2
```

**Comment by Tiryoh** [comment]
Adding `#include <linux/uaccess.h>` to `rtmouse.c` solved the problem on Ubuntu 18.04(Linux Kernel Version : 4.15.0).

diff:
```
ubuntu@ubuntu:~/RaspberryPiMouse$ git diff
diff --git a/src/drivers/rtmouse.c b/src/drivers/rtmouse.c
index d0a65d5..e40971c 100644
--- a/src/drivers/rtmouse.c
+++ b/src/drivers/rtmouse.c
@@ -47,6 +47,7 @@

 #include <asm/delay.h>
 #include <asm/uaccess.h>
+#include <linux/uaccess.h>

 #define RASPBERRYPI2
 #undef RASPBERRYPI1
```

build log:
```
ubuntu@ubuntu:~$ ~/RaspberryPiMouse/utils/build_install.ubuntu14.bash
#!/bin/bash -vxe

dir=$(dirname $0)/../
++ dirname /home/ubuntu/RaspberryPiMouse/utils/build_install.ubuntu14.bash
+ dir=/home/ubuntu/RaspberryPiMouse/utils/../

cd $dir/src/drivers/
+ cd /home/ubuntu/RaspberryPiMouse/utils/..//src/drivers/
rm Makefile
+ rm Makefile
ln -s Makefile.ubuntu14 Makefile
+ ln -s Makefile.ubuntu14 Makefile
make clean
+ make clean
make -C /usr/src/linux-headers-`uname -r` M=`pwd` V=1 clean
make[1]: Entering directory '/usr/src/linux-headers-4.15.0-1010-raspi2'
make -f ./scripts/Makefile.clean obj=/home/ubuntu/RaspberryPiMouse/src/drivers
  rm -rf /home/ubuntu/RaspberryPiMouse/src/drivers/.tmp_versions
  rm -f /home/ubuntu/RaspberryPiMouse/src/drivers/Module.symvers
make[1]: Leaving directory '/usr/src/linux-headers-4.15.0-1010-raspi2'
make
+ make
make -C /usr/src/linux-headers-`uname -r` M=`pwd` V=1 modules
make[1]: Entering directory '/usr/src/linux-headers-4.15.0-1010-raspi2'
test -e include/generated/autoconf.h -a -e include/config/auto.conf || (		\
echo >&2;							\
echo >&2 "  ERROR: Kernel configuration is invalid.";		\
echo >&2 "         include/generated/autoconf.h or include/config/auto.conf are missing.";\
echo >&2 "         Run 'make oldconfig && make prepare' on kernel src to fix it.";	\
echo >&2 ;							\
/bin/false)
mkdir -p /home/ubuntu/RaspberryPiMouse/src/drivers/.tmp_versions ; rm -f /home/ubuntu/RaspberryPiMouse/src/drivers/.tmp_versions/*
make -f ./scripts/Makefile.build obj=/home/ubuntu/RaspberryPiMouse/src/drivers
  gcc -Wp,-MD,/home/ubuntu/RaspberryPiMouse/src/drivers/.rtmouse.o.d  -nostdinc -isystem /usr/lib/gcc/arm-linux-gnueabihf/7/include  -I./arch/arm/include -I./arch/arm/include/generated  -I./include -I./arch/arm/include/uapi -I./arch/arm/include/generated/uapi -I./include/uapi -I./include/generated/uapi -include ./include/linux/kconfig.h -Iubuntu/include  -D__KERNEL__ -mlittle-endian -Wall -Wundef -Wstrict-prototypes -Wno-trigraphs -fno-strict-aliasing -fno-common -fshort-wchar -Werror-implicit-function-declaration -Wno-format-security -std=gnu89 -fno-PIE -fno-dwarf2-cfi-asm -fno-omit-frame-pointer -mapcs -mno-sched-prolog -fno-ipa-sra -mabi=aapcs-linux -mno-thumb-interwork -mfpu=vfp -funwind-tables -marm -D__LINUX_ARM_ARCH__=7 -march=armv7-a -msoft-float -Uarm -fno-delete-null-pointer-checks -Wno-frame-address -Wno-format-truncation -Wno-format-overflow -Wno-int-in-bool-context -O2 --param=allow-store-data-races=0 -DCC_HAVE_ASM_GOTO -Wframe-larger-than=1024 -fstack-protector-strong -Wno-unused-but-set-variable -Wno-unused-const-variable -fno-omit-frame-pointer -fno-optimize-sibling-calls -fno-var-tracking-assignments -pg -Wdeclaration-after-statement -Wno-pointer-sign -fno-strict-overflow -fno-merge-all-constants -fmerge-constants -fno-stack-check -fconserve-stack -Werror=implicit-int -Werror=strict-prototypes -Werror=date-time -Werror=incompatible-pointer-types -Werror=designated-init  -DMODULE  -DKBUILD_BASENAME='"rtmouse"'  -DKBUILD_MODNAME='"rtmouse"' -c -o /home/ubuntu/RaspberryPiMouse/src/drivers/.tmp_rtmouse.o /home/ubuntu/RaspberryPiMouse/src/drivers/rtmouse.c
  if [ "-pg" = "-pg" ]; then if [ /home/ubuntu/RaspberryPiMouse/src/drivers/rtmouse.o != "scripts/mod/empty.o" ]; then ./scripts/recordmcount  "/home/ubuntu/RaspberryPiMouse/src/drivers/rtmouse.o"; fi; fi;
(cat /dev/null;   echo kernel//home/ubuntu/RaspberryPiMouse/src/drivers/rtmouse.ko;) > /home/ubuntu/RaspberryPiMouse/src/drivers/modules.order
make -f ./scripts/Makefile.modpost
  find /home/ubuntu/RaspberryPiMouse/src/drivers/.tmp_versions -name '*.mod' | xargs -r grep -h '\.ko$' | sort -u | sed 's/\.ko$/.o/' | scripts/mod/modpost -m -a -i ./Module.symvers -I /home/ubuntu/RaspberryPiMouse/src/drivers/Module.symvers  -o /home/ubuntu/RaspberryPiMouse/src/drivers/Module.symvers -S  -w  -s -T -
  gcc -Wp,-MD,/home/ubuntu/RaspberryPiMouse/src/drivers/.rtmouse.mod.o.d  -nostdinc -isystem /usr/lib/gcc/arm-linux-gnueabihf/7/include  -I./arch/arm/include -I./arch/arm/include/generated  -I./include -I./arch/arm/include/uapi -I./arch/arm/include/generated/uapi -I./include/uapi -I./include/generated/uapi -include ./include/linux/kconfig.h -Iubuntu/include  -D__KERNEL__ -mlittle-endian -Wall -Wundef -Wstrict-prototypes -Wno-trigraphs -fno-strict-aliasing -fno-common -fshort-wchar -Werror-implicit-function-declaration -Wno-format-security -std=gnu89 -fno-PIE -fno-dwarf2-cfi-asm -fno-omit-frame-pointer -mapcs -mno-sched-prolog -fno-ipa-sra -mabi=aapcs-linux -mno-thumb-interwork -mfpu=vfp -funwind-tables -marm -D__LINUX_ARM_ARCH__=7 -march=armv7-a -msoft-float -Uarm -fno-delete-null-pointer-checks -Wno-frame-address -Wno-format-truncation -Wno-format-overflow -Wno-int-in-bool-context -O2 --param=allow-store-data-races=0 -DCC_HAVE_ASM_GOTO -Wframe-larger-than=1024 -fstack-protector-strong -Wno-unused-but-set-variable -Wno-unused-const-variable -fno-omit-frame-pointer -fno-optimize-sibling-calls -fno-var-tracking-assignments -pg -Wdeclaration-after-statement -Wno-pointer-sign -fno-strict-overflow -fno-merge-all-constants -fmerge-constants -fno-stack-check -fconserve-stack -Werror=implicit-int -Werror=strict-prototypes -Werror=date-time -Werror=incompatible-pointer-types -Werror=designated-init  -DKBUILD_BASENAME='"rtmouse.mod"'  -DKBUILD_MODNAME='"rtmouse"' -DMODULE  -c -o /home/ubuntu/RaspberryPiMouse/src/drivers/rtmouse.mod.o /home/ubuntu/RaspberryPiMouse/src/drivers/rtmouse.mod.c
  ld -EL -r  -T ./scripts/module-common.lds --build-id  -o /home/ubuntu/RaspberryPiMouse/src/drivers/rtmouse.ko /home/ubuntu/RaspberryPiMouse/src/drivers/rtmouse.o /home/ubuntu/RaspberryPiMouse/src/drivers/rtmouse.mod.o ;  true
make[1]: Leaving directory '/usr/src/linux-headers-4.15.0-1010-raspi2'
sudo insmod rtmouse.ko
+ sudo insmod rtmouse.ko
sleep 1
+ sleep 1
sudo chmod 666 /dev/rt*
+ sudo chmod 666 /dev/rtbuzzer0 /dev/rtcounter0 /dev/rtcounter_l0 /dev/rtcounter_r0 /dev/rtled0 /dev/rtled1 /dev/rtled2 /dev/rtled3 /dev/rtlightsensor0 /dev/rtmotor0 /dev/rtmotor_raw_l0 /dev/rtmotor_raw_r0 /dev/rtmotoren0 /dev/rtswitch0 /dev/rtswitch1 /dev/rtswitch2
echo 0 > /dev/rtmotoren0
+ echo 0
```

**Comment by Tiryoh** [comment]
`asm/uaccess.h` does not have `copy_to_user` on kernel > 4.11.
https://github.com/raspberrypi/linux/commit/d597580d373774b1bdab84b3d26ff0b55162b916
https://github.com/raspberrypi/linux/commit/701cac61d0250912b89cbc28589969530179099a#diff-073429958395ac01bbdf79a3face785d
